### PR TITLE
use precreated labeled namespace for workspaces

### PR DESCRIPTION
**Pull request by sparkoo**
Signed-off-by: Michal Vala <mvala@redhat.com>

### What does this PR do?
Allow using precreated namespaces labeled with proper labels. The labels are configurable with `che.infra.kubernetes.namespace.labels` property.

### Documentation PR
https://github.com/eclipse/che-docs/pull/1643

TODO:
 - [x] Make labeled namespace as a valid namespace for creating new workspace (TODO in KubernetesNamespaceFactory.java)
 - [x] Javadocs
 - [x] Tests
 - [x] Documentation

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/17842

### How to test this PR?

1. create namespace with labels `app.kubernetes.io/part-of=che.eclipse.org,app.kubernetes.io/component=workspace,che.eclipse.org/username=<username>` and make sure your user has access to it (`admin` role)
  1. On OpenShift with OAuth, it should be enough just to create RoleBinding for the user and `admin` role in pre-created project.
  1. On Kubernetes, you will also need to add list namespaces clusterrole permission to `che` serviceaccount.
1. login to Che with the user
1. go to create new namespace page
1. you should see the namespace name in the dropdown "Namespace"
1. when you use factory to create a namespace, this labeled namespace will be used
1. when you label more namespaces, all will be shown in dashboard. If you use factory with more labeled namespaces, the first one will be used (should be alphabetical order, but not guaranteed)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 1]
```diff
@@ -505,6 +557,30 @@ public String evaluateNamespaceName(NamespaceResolutionContext resolutionCtx)
...
+        .namespaces()
+        .withLabels(evalLabels(username))
+        .list()
+        .getItems()
+        .stream()
+        .map(n -> n.getMetadata().getName())
+        .collect(Collectors.toList());
+  }
+
+  protected Map<String, String> evalLabels(String username) {
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: `evaluateLabels`?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 2]
```diff
@@ -135,9 +140,21 @@ public KubernetesNamespaceFactory(
...
+  private Map<String, String> evalNamespaceLabels(String namespaceLabels) {
+    if (namespaceLabels == null) {
+      return emptyMap();
+    }
+
+    //noinspection UnstableApiUsage
+    Map<String, String> labelsMap =
+        Splitter.on(",").withKeyValueSeparator("=").split(namespaceLabels);
+
+    if (Strings.countOccurrencesOf(String.join("", labelsMap.values()), "%s") != 1) {
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: I wanted to limit it somehow. As I'm thinking about it, it really feels like artificial limitation. Removed and waiting what crazy scenarios will users create with that :))

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 3]
```diff
@@ -297,6 +297,9 @@ che.infra.kubernetes.namespace.creation_allowed=true
 # Is used by OpenShift infra as well to specify Project
 che.infra.kubernetes.namespace.default=<username>-che
 
+# List of labels to prepare Namespaces/Projects for the Che users.
```
Comment (sparkoo) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: better doc, mention placeholders

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 4]
```diff
@@ -467,13 +477,49 @@ public String evaluateLegacyNamespaceName(NamespaceResolutionContext resolutionC
    */
   public String evaluateNamespaceName(NamespaceResolutionContext resolutionCtx)
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:495]: update javadoc

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 5]
```diff
@@ -467,13 +477,49 @@ public String evaluateLegacyNamespaceName(NamespaceResolutionContext resolutionC
...
       throws InfrastructureException {
-    String namespace;
+    return findFirstLabeledNamespace(resolutionCtx)
+        .or(() -> findStoredNamespace(resolutionCtx))
+        .orElse(evalDefaultNamespace(resolutionCtx));
+  }
+
+  private Optional<String> findFirstLabeledNamespace(NamespaceResolutionContext resolutionCtx)
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:511]: javadoc

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 6]
```diff
@@ -467,13 +477,49 @@ public String evaluateLegacyNamespaceName(NamespaceResolutionContext resolutionC
...
+            labeledNamespaces.toString(),
+            foundNamespace);
+      }
+      return Optional.of(foundNamespace);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private Optional<String> findStoredNamespace(NamespaceResolutionContext resolutionCtx) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:542]: javadoc

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 7]
```diff
@@ -467,13 +477,49 @@ public String evaluateLegacyNamespaceName(NamespaceResolutionContext resolutionC
...
-      namespace = evalPlaceholders(defaultNamespaceName, resolutionCtx);
+    if (storedNamespace.isPresent() && isStoredTemplateValid(storedNamespace.get().second)) {
+      return Optional.of(storedNamespace.get().first);
     } else {
-      namespace = storedNamespace.get().first;
+      return Optional.empty();
     }
+  }
+
+  private String evalDefaultNamespace(NamespaceResolutionContext resolutionCtx)
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:557]: javadoc

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 8]
```diff
@@ -505,6 +551,29 @@ public String evaluateNamespaceName(NamespaceResolutionContext resolutionCtx)
...
+        .namespaces()
+        .withLabels(evaluateLabels(namespaceCtx))
+        .list()
+        .getItems()
+        .stream()
+        .map(this::asNamespaceMeta)
+        .collect(Collectors.toList());
+  }
+
+  protected Map<String, String> evaluateLabels(NamespaceResolutionContext namespaceCtx) {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: - javadoc
 - rename `evaluateLabelsPlaceholders`

**Review by sparkoo — COMMENTED** [review]
I've changed the labels property to use placeholders, because without them, we would not be able to be really flexible with label names.

**Comment by artaleks9** [comment]
[crw-ci-test]

**Comment by skabashnyuk** [comment]
looks 'ok' in general

**Comment by sparkoo** [comment]
[crw-ci-test]

**Comment by sparkoo** [comment]
@metlos @skabashnyuk @mshaposhnik ready to review

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**